base/initdefs.jl

```diff
@@ -374,7 +375,7 @@ calls `exit(n)`, then Julia will exit with the exit code corresponding to the
 last called exit hook that calls `exit(n)`. (Because exit hooks are called in
 LIFO order, "last called" is equivalent to "first registered".)
 """
-atexit(f::Function) = (pushfirst!(atexit_hooks, f); nothing)
+atexit(f::Function) = Base.@lock _atexit_hooks_lock (pushfirst!(atexit_hooks, f); nothing)
 
 function _atexit(exitcode::Cint)
     while !isempty(atexit_hooks)
```

Comment:
Actually, maybe we should lock here too... Is there any chance that other Julia Threads can still be running during `_atexit()`? I guess it can't hurt to grab the lock here... Lemme push that up too.

Is it still safe to call into the runtime for things like locks inside `_atexit()`? For this we might need help from @vtjnash or someone similarly familiar?

Reply:
It [seems](https://stackoverflow.com/questions/39655868/what-does-the-posix-standard-say-about-thread-stacks-in-atexit-handlers-what) threads may or may not have stopped running.

Don't think it's safe/possible to lock in exit handler context either. It's an improbable race too, maybe?

I'd leave it, but maybe @vtjnash has thoughts on this?

Reply:
We can always open a follow-up issue to cover this question.

Reply:
K, filed: https://github.com/JuliaLang/julia/issues/49841 as followup.